File: src/main/java/io/github/numichi/reactive/logger/MDC.java
package io.github.numichi.reactive.logger;

import java.util.Map;
import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;

public class MDC extends ConcurrentHashMap<String, String> {
    private final String assignedContextKey;
    
    public MDC() {
        this(null, null);
    }
    
    public MDC(String assignedContextKey) {
        this(assignedContextKey, null);
    }
    
    public MDC(Map<String, String> mdcMap) {
        this(null, mdcMap);
    }
    
    public MDC(String assignedContextKey, Map<String, String> mdcMap) {
        if (mdcMap != null) {
            this.putAll(mdcMap);
        }
    
        if (assignedContextKey != null) {
            this.assignedContextKey = assignedContextKey;
        } else {
            this.assignedContextKey = DefaultValues.getInstance().getDefaultReactorContextMdcKey();
        }
    }
    
    public String getAssignedContextKey() {
        return assignedContextKey;
    }
    
    @Deprecated
    public Map<String, String> asMap() {
        return this;
    }
    
    @Deprecated
    public String getContextKey() {
        return assignedContextKey;
    }
    
    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof MDC)) {
            return false;
        }
        if (!super.equals(o)) {
            return false;
        }
        MDC mdc = (MDC) o;
        return getAssignedContextKey().equals(mdc.getAssignedContextKey());
    }
    
    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), getAssignedContextKey());
    }
}

File: src/main/java/io/github/numichi/reactive/logger/MDC.kt
package io.github.numichi.reactive.logger

import java.util.concurrent.ConcurrentHashMap

open class MDC(contextKey: String?, mdcMap: Map<String, String>) : ConcurrentHashMap<String, String>() {
    val contextKey: String

    constructor() : this(null, mapOf())
    constructor(contextKey: String) : this(contextKey, mapOf())
    constructor(mdcMap: Map<String, String>) : this(null, mdcMap)
    constructor(mdc: MDC) : this(mdc.contextKey, mdc)

    init {
        this.putAll(mdcMap)
        this.contextKey = contextKey ?: Configuration.defaultReactorContextMdcKey
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is MDC) return false
        if (!super.equals(other)) return false
        if (contextKey != other.contextKey) return false
        return true
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + contextKey.hashCode()
        return result
    }
}